File: common/src/main/java/com/lunatech/kafka/common/java/StockMessage.java
package com.lunatech.kafka.common.java;

import java.math.BigDecimal;
import java.math.MathContext;

public class StockMessage {
    private String symbol;
    private BigDecimal value;

    public StockMessage(String symbol, BigDecimal value) {
        this.symbol = symbol;
        this.value = value.round(new MathContext(8));
    }

    public String getSymbol() {
        return symbol;
    }

    public BigDecimal getValue() {
        return value;
    }
}

File: common/src/main/scala/com/lunatech/kafka/common/scala/StockMessage.scala
package com.lunatech.kafka.common.scala

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule

import scala.math.BigDecimal.RoundingMode

case class StockMessage private (symbol: String, value: BigDecimal)

object StockMessage {
  private val mapper = new ObjectMapper()
  mapper.registerModule(DefaultScalaModule)

  def apply(symbol: String, value: BigDecimal): StockMessage =
    new StockMessage(symbol, value.setScale(4, RoundingMode.HALF_EVEN))

  def serialize(m: StockMessage): Array[Byte] = mapper.writeValueAsBytes(m)

  def deserialize(m: Array[Byte]): StockMessage =
    mapper.readValue(m, classOf[StockMessage])
}
